File: Server/src/main/java/core/game/node/entity/skill/construction/ConstructionInterface.java
package core.game.node.entity.skill.construction;


import core.cache.def.impl.ItemDefinition;
import core.game.component.Component;
import core.game.component.ComponentDefinition;
import core.game.component.ComponentPlugin;
import core.game.node.entity.player.Player;
import core.game.node.entity.skill.Skills;
import core.game.node.scenery.Scenery;
import core.plugin.Initializable;
import core.plugin.Plugin;
import rs09.game.node.entity.skill.construction.Hotspot;
import rs09.game.system.SystemLogger;

/**
 * Handles the creating of a decoration object.
 * @author Emperor
 *
 */
@Initializable
public final class ConstructionInterface extends ComponentPlugin {

	@Override
	public Plugin<Object> newInstance(Object arg) {
		ComponentDefinition.put(396, this);
		ComponentDefinition.put(398, this);
		ComponentDefinition.put(402, this);
		return this;
	}

	@Override
	public boolean handle(Player player, Component component, int opcode, int button, int slot, int itemId) {
		switch (component.getId()) {
			case 396:
				switch (button) {
					case 132:
						player.getInterfaceManager().close();
						Hotspot hotspot = player.getAttribute("con:hotspot");
						Scenery object = player.getAttribute("con:hsobject");
						if (hotspot == null || object == null) {
							SystemLogger.logErr("Failed building decoration " + hotspot + " : " + object);
							break;
						}
						slot = ((slot % 2 != 0) ? 4 : 0) + (slot >> 1);
						if (slot >= hotspot.getHotspot().getDecorations().length) {
							SystemLogger.logErr("Failed building decoration " + slot + "/" + hotspot.getHotspot().getDecorations().length);
							break;
						}
						boolean debug = player.isStaff();
						Decoration deco = hotspot.getHotspot().getDecorations()[slot];
						if (!debug) {
							if (player.getSkills().getLevel(Skills.CONSTRUCTION) < deco.getLevel()) {
								player.sendMessage("You need to have a Construction level of " + deco.getLevel() + " to build that.");
								return true;
							}
							if (!player.getInventory().containsItems(deco.getItems())) {
								player.sendMessage("You don't have the right materials.");
								return true;
							}
							for (int tool : deco.getTools()) {
								if (tool == BuildingUtils.WATERING_CAN) {
									boolean hasWateringCan = false;
									for (int i = 0; i < 8; i++) {
										if (player.getInventory().contains(tool - i, 1)) {
											hasWateringCan = true;
											break;
										}
									}
									if (!hasWateringCan) {
										player.sendMessage("You need a watering can to plant this.");
										return true;
									}
									continue;
								}
								if (!player.getInventory().contains(tool, 1)) {
									player.sendMessage("You need a " + ItemDefinition.forId(tool).getName() + " to build this.");
									return true;
								}
							}
						}
						BuildingUtils.buildDecoration(player, hotspot, deco, object);
						return true;
				}
				break;
			case 398:
				switch (button) {
					case 14:
						player.getHouseManager().toggleBuildingMode(player, true);
						return true;
					case 1:
						player.getHouseManager().toggleBuildingMode(player, false);
						return true;
					case 15:
						player.getHouseManager().expelGuests(player);
						return true;
					case 13:
						if (!player.getHouseManager().isInHouse(player)) {
							player.sendMessage("You can't do this outside of your house.");
							return true;
						}
						HouseManager.leave(player);
						return true;
				}
				break;
			case 402:
				int index = button - 160;
				SystemLogger.logErr("BuildRoom Interface Index: " + index);
				if (index > -1 && index < RoomProperties.values().length) {
					player.getDialogueInterpreter().open("con:room", RoomProperties.values()[index]);
					return true;
				}
				break;
		}
		return false;
	}

}

File: Server/src/main/content/global/skill/construction/ConstructionInterface.kt
package content.global.skill.construction

import content.global.skill.construction.decoration.workshop.WorkbenchListeners
import core.api.*
import core.cache.def.impl.ItemDefinition
import core.game.interaction.InterfaceListener
import core.game.node.entity.skill.Skills
import core.game.node.scenery.Scenery
import core.tools.Log
import org.rs09.consts.Components
import org.rs09.consts.Scenery as SceneryObj

/**
 * Handles three interfaces related to the Construction skill, namely those which allow a player to build decorations,
 * toggle building mode/expel guests, and select a room to add to the POH.
 * @author Bishop
 */

class ConstructionInterface : InterfaceListener {

    companion object {
        const val ATTRIBUTE_HOTSPOT                  = "con:hotspot"
        const val ATTRIBUTE_HOTSPOT_OBJ              = "con:hsobject"
        private const val DECO_BUTTON_BUILD          = 132
        private const val HOUSE_BUTTON_BUILDMODE_ON  = 14
        private const val HOUSE_BUTTON_BUILDMODE_OFF = 1
        private const val HOUSE_BUTTON_EXPEL_GUESTS  = 15
        private const val HOUSE_BUTTON_LEAVE_HOUSE   = 13

        // Register furniture here when it is only obtainable by upgrading, to send the correct rejection message.
        private val upgradeOnly = arrayOf(
            Decoration.WORKBENCH_WITH_VICE,
            Decoration.WORKBENCH_WITH_LATHE,
        )
    }

    override fun defineInterfaceListeners() {
        /**
         * Handles the interface for selecting a decoration to build.
         */
        on(Components.POH_DECORATION_396) { player, _, _, buttonID, slot, _ ->
            when (buttonID) {
                DECO_BUTTON_BUILD -> {
                    val hotspot = getAttribute<Hotspot?>(player, ATTRIBUTE_HOTSPOT, null)
                    val hotspotObj = getAttribute<Scenery?>(player, ATTRIBUTE_HOTSPOT_OBJ, null)
                    val flatpackMode = getAttribute(player, WorkbenchListeners.ATTRIBUTE_FLATPACK_MODE, false)
                    if (!flatpackMode) {
                        closeInterface(player)
                    }
                    if ((hotspot == null || hotspotObj == null) && !flatpackMode) {
                        log(this.javaClass, Log.ERR, "Failed building decoration $hotspot : $hotspotObj")
                        return@on false
                    }
                    // Translates slot data to index data, as slots are not in order of level requirement in the interface, but decoration data is organized that way
                    val decoIndex = (if (slot % 2 != 0) 4 else 0) + (slot shr 1)
                    val buildHotspot = if (flatpackMode) {
                        getAttribute<BuildHotspot?>(player, WorkbenchListeners.ATTRIBUTE_WORKBENCH_SELECTION, null)?: return@on false
                    } else {
                        hotspot!!.hotspot
                    }
                    if (decoIndex >= buildHotspot.decorations.size) {
                        return@on false
                    }
                    val deco = buildHotspot.decorations[decoIndex]
                    if (!player.isAdmin) {
                        if (getDynLevel(player, Skills.CONSTRUCTION) < deco.level) {
                            closeInterface(player)
                            sendMessage(player, "You need to have a Construction level of ${deco.level} to build that.")
                            return@on true
                        }
                        if (deco in upgradeOnly) {
                            closeInterface(player)
                            sendMessage(player, "That can only be built by upgrading the previous piece of furniture.")
                            return@on true
                        }
                        if (!player.inventory.containsItems(*deco.items)) { // No clean ContentAPI substitute for this
                            closeInterface(player)
                            sendMessage(player, "You don't have the right materials.")
                            return@on true
                        }
                        for (tool in deco.tools) {
                            if (tool == BuildingUtils.WATERING_CAN) {
                                var hasWaterInCan = false
                                for (i in 0..8) {
                                    if (inInventory(player, tool - i)) {
                                        hasWaterInCan = true
                                        break
                                    }
                                }
                                if (!hasWaterInCan) {
                                    sendMessage(player, "You need a watering can to plant this.")
                                    return@on true
                                }
                                continue
                            }
                            if (!inInventory(player, tool)) {
                                closeInterface(player)
                                sendMessage(player, "You need a ${ItemDefinition.forId(tool).name} to build this.")
                                return@on true
                            }
                        }
                        if (flatpackMode && WorkbenchListeners.getBenchLevel(getAttribute(player, WorkbenchListeners.ATTRIBUTE_WORKBENCH_ID, SceneryObj.WORKBENCH_13704)) < deco.level) {
                            closeInterface(player)
                            sendMessage(player, "You cannot make this with this workbench.") // TODO: Find authentic text
                            return@on true
                        }
                    }
                    if (flatpackMode) {
                        WorkbenchListeners.produceFlatpack(player, deco)
                    } else {
                        BuildingUtils.buildDecoration(player, hotspot, deco, hotspotObj)
                    }
                    return@on true
                }
                else -> return@on false
            }
        }

        /**
         * Handles the interface for managing building mode and POH guests.
         */
        on(Components.POH_HOUSE_OPTIONS_398) { player, _, _, buttonID, _, _ ->
            when (buttonID) {
                HOUSE_BUTTON_BUILDMODE_ON -> {
                    player.houseManager.toggleBuildingMode(player, true)
                    return@on true
                }
                HOUSE_BUTTON_BUILDMODE_OFF -> {
                    player.houseManager.toggleBuildingMode(player, false)
                    return@on true
                }
                HOUSE_BUTTON_EXPEL_GUESTS -> {
                    player.houseManager.expelGuests(player)
                    return@on true
                }
                HOUSE_BUTTON_LEAVE_HOUSE -> {
                    if (!player.houseManager.isInHouse(player)) {
                        sendMessage(player, "You can't do this outside of your house.")
                        return@on true
                    }
                    HouseManager.leave(player)
                    return@on true
                }
                else -> return@on false
            }
        }

        /**
         * Handles the interface for selecting a new room to add to a POH.
         */
        on(Components.POH_ROOMS_402) { player, _, _, buttonID, _, _ ->
            val index = buttonID - 160
            if (index > -1 && index < RoomProperties.values().size) {
                player.dialogueInterpreter.open("con:room", RoomProperties.values()[index]) // ContentAPI impl won't work for this
            }
            return@on true
        }
    }
}